11_FFTBloom/app_resources/image_fft_first_axis.hlsl

```diff
@@ -0,0 +1,156 @@
+uint32_t3 glsl::gl_WorkGroupSize() { return uint32_t3(_NBL_HLSL_WORKGROUP_SIZE_, 1, 1); }
+
+struct SharedMemoryAccessor
+{
+	void set(uint32_t idx, uint32_t value)
+	{
+		sharedmem[idx] = value;
+	}
+
+	void get(uint32_t idx, NBL_REF_ARG(uint32_t) value)
+	{
+		value = sharedmem[idx];
+	}
+
+	void workgroupExecutionAndMemoryBarrier()
+	{
+		glsl::barrier();
+	}
+
+};
+
+// ---------------------------------------------------- Utils ---------------------------------------------------------
+
+// After first FFT we only store half of a column, so the offset per column is half the image side length
+uint32_t colMajorOffset(uint32_t x, uint32_t y)
+{
+	return x * (FFT_LENGTH / 2) | y;
+}
+
+// We store half of a column, but unlike the kernel we only store as many columns as there actually are in the image
+// We launch one workgroup every two columns in the image, and we only write (per channel) one column per column in the image
+// This is unlike the kernel, in which we always upsample it to nearest PoT
+// So there are `2 * glsl::gl_NumWorkGroups().x` columns per channel, each of size `FFT_LENGTH / 2` (we only store half) and the 2's cancel out
+uint64_t getChannelStartAddress(uint32_t channel)
+{
+	return pushConstants.colMajorBufferAddress + channel * glsl::gl_NumWorkGroups().x * FFT_LENGTH * sizeof(complex_t<scalar_t>);
+}
+
+
+// -------------------------------------------- FIRST AXIS FFT ------------------------------------------------------------------
+
+// Each Workgroup computes the FFT along two consecutive vertical scanlines (fixed x for the whole Workgroup) so we use `2 * gl_WorkGroupID().x, 2 * gl_WorkGroupID().x + 1` 
+// to get the x coordinates for each of the consecutive lines. This time we launch `inputImageSize.x / 2` workgroups 
+
+struct PreloadedFirstAxisAccessor : PreloadedAccessorBase<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_, scalar_t> 
+{
+
+	void preload(uint32_t channel)
+	{
+		float32_t2 inputImageSize;
+		texture.GetDimensions(inputImageSize.x, inputImageSize.y);
+		float32_t2 normalizedCoordsFirstLine, normalizedCoordsSecondLine;
+		normalizedCoordsFirstLine.x = (float32_t(2 * glsl::gl_WorkGroupID().x) + 0.5f) / inputImageSize.x;
+		normalizedCoordsSecondLine.x = (float32_t(2 * glsl::gl_WorkGroupID().x + 1) + 0.5f) / inputImageSize.x;
+
+		// Remember to add padding before and after - we will be sampling the original image mirrored at the borders - this avoids loss of brightness at the edges
+		const uint32_t padding = uint32_t(FFT_LENGTH - inputImageSize.y) >> 1;
+
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex++)
+		{
+			// Index computation here is easier than FFT since the stride is fixed at _NBL_HLSL_WORKGROUP_SIZE_
+			const uint32_t index = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+			normalizedCoordsFirstLine.y = (float32_t(index) - padding + 0.5f) / inputImageSize.y;
```

Comment:
hoist the computation of `(workgroup::SubgroupCOntiguousIndex()-padding+0.5)/inputImage.y` then just `normalizedCoordsFirstLine.y += WorkgroupSize/imageHeightRcp.y;` (obviously have that be one precomputed as one push constant var.

Reply:
also `(workgroup::SubgroupCOntiguousIndex()-padding+0.5)/inputImage.y` can be expressed as `workgroup::SubgroupCOntiguousIndex()*pc.mulConst+pc.addConst`